### ISSUE TITLE
Ansi Console integration needs polishing

### ISSUE THREAD

**Issue by iloveeclipse** [eclipse-platform/eclipse.platform.debug#57]
Follow up on https://github.com/eclipse-platform/eclipse.platform.debug/pull/50.

I see multiple issues with the new contribution that need to be fixed.

1. There are two new keybindings that use "Ctrl+Alt+" combination - that "Ctrl+Alt" shouldn't be used because it is mapped differently (as "AltGr") on international keyboards on Windows. Either change to "Ctrl+Shift+" or simply disable to avoid problems.
2. There are three new buttons added to Console view that don't need to be there at all:  
![image](https://user-images.githubusercontent.com/964108/183419025-1c235642-6760-4087-8d19-ba0b68d386a9.png)
  a) The "Options" button should be contributed as "standard" view options button and "Ansi Support Preferences..." menu there, along with (global) "Console Preferences..." that is missing today.
  b) The "Enable/Disable ANSI Console" isn't needed to be on global toolbar -> it should go to the "standard" view options menu
  c) The "Copy text with escapes" is most likely even less common use case a "Copy" - but there is no button for "Copy", so there new button should be also moved to the "standard" location - right click into the console, next to "Copy" menu. 
  d)  "Copy text with escapes" should be capitalized properly  "Copy Text With Escapes" 
3. Preference page mentions "Preferences for Ansi Console". This should be completely removed - there is no "Ansi Console" anymore, it is "Ansi Support", but that is already the preference topic. 
4. The next item "Plugin enabled" doesn't make sense anymore. It should be renamed to "Disable Ansi support"
![image](https://user-images.githubusercontent.com/964108/183419571-2170fd50-856b-4f40-9d02-2b7a2cd92f32.png)
5. Context menu "Copy" in the console doesn't show now the "Ctrl+C" binding because it is "stolen" by Ansi Console code "Copy without escapes".
6. "Ansi" in Java code is OK, but not in UI, because it is ASCII, ANSI, etc not Ascii, Ansii etc.

@mihnita : It would be nice f the points above would be addressed by you soon.

**Comment by iloveeclipse** [comment]
@laeubi : the issues above aren't enhancements, they are **already in the code**. :-)

**Comment by laeubi** [comment]
@iloveeclipse maybe wen need another label then, bug is described as "something isn't working", as everything you describe here does work I won't say it is a bug ... or do you find something that is not working at all (e.g. disabling the console support or copy with escapes?)

**Comment by iloveeclipse** [comment]
It is a common misunderstanding that usability issues or inconsistencies in the UI's aren't bugs. They are. 

"Bug" does not always mean something isn't working - if it might work, but not for all, or not always as expected, or surprises user or needs longer time to learn etc.

So here we have clearly not "everything is broken" case but just "standard" usability issues.

**Comment by laeubi** [comment]
> It is a common misunderstanding that usability issues or inconsistencies in the UI's aren't bugs. They are.

I think it is worth to distinguish this more, as all of these are clearly *issues* to be solved, but at lest for me a *bug* is something that (unintentionally) is not working as designed and for sure no one would add a bug by purpose.

So all things mentioned here, are there *by purpose* but still could be *improved* (thus the enhancement label) because all you mentioned sounds a good approach, but maybe other people might have different feelings *where* to place a button, but if the button do *not work* at all this is not really a matter of usability/improvement.

Anyways it seems not that relevant, I just find it a bit nicer in the overall communication to not say "something is not working" over "something should be improved", but this as well might be a personal feeling and might vary...

**Comment by iloveeclipse** [comment]
> I just find it a bit nicer in the overall communication to not say "something is not working" 

Where did you saw that?

> over "something should be improved"

You probably didn't noticed the title "Ansi Console integration needs polishing" ?

**Comment by laeubi** [comment]
> Where did you saw that?

![grafik](https://user-images.githubusercontent.com/1331477/183427782-d84cf7d1-8b66-4b24-8df0-93533b3adabc.png)

> You probably didn't noticed the title "Ansi Console integration needs polishing" ?

I noticed that and thus though it might be more an enhancement, seems I overlooked the usability label ...

**Comment by iloveeclipse** [comment]
Is this better now: ![image](https://user-images.githubusercontent.com/964108/183428474-2ad6ff8b-9992-426f-a6e6-78b9ac60dab6.png)?

**Comment by laeubi** [comment]
Much better now ;-)

**Comment by iloveeclipse** [comment]
Keybinding seems to be problematic: I believe the context is not properly set somewhere, it needs to be investigated why: I get this error in "Java Stack Trace Console" if re-mapping to "Ctrl+Shift+":
```
!ENTRY org.eclipse.jface 2 0 2022-08-08 16:11:06.997
!MESSAGE Keybinding conflicts occurred.  They may interfere with normal accelerator operation.
!SUBENTRY 1 org.eclipse.jface 2 0 2022-08-08 16:11:06.997
!MESSAGE A conflict occurred for CTRL+SHIFT+C:
Binding(CTRL+SHIFT+C,
	ParameterizedCommand(Command(org.eclipse.jdt.ui.edit.text.java.toggle.comment,Toggle Comment,
		Toggle comment the selected lines,
		Category(org.eclipse.jdt.ui.category.source,Source,Java Source Actions,true),
		WorkbenchHandlerServiceHandler("org.eclipse.jdt.ui.edit.text.java.toggle.comment"),
		,,true),null),
	org.eclipse.ui.defaultAcceleratorConfiguration,
	org.eclipse.jdt.ui.javaEditorScope,,,system)
Binding(CTRL+SHIFT+C,
	ParameterizedCommand(Command(AnsiConsole.command.copy_with_escapes,Copy Text With Escapes,
		Copy the console content to clipboard, including the escape sequences,
		Category(AnsiConsole.command.categoryid,Ansi Support Commands,null,true),
		WorkbenchHandlerServiceHandler("AnsiConsole.command.copy_with_escapes"),
		,,true),null),
	org.eclipse.ui.defaultAcceleratorConfiguration,
	org.eclipse.ui.console.ConsoleView,,,system)
!SUBENTRY 1 org.eclipse.jface 2 0 2022-08-08 16:11:06.998
!MESSAGE A conflict occurred for CTRL+SHIFT+INSERT:
Binding(CTRL+SHIFT+INSERT,
	ParameterizedCommand(Command(AnsiConsole.command.copy_with_escapes,Copy Text With Escapes,
		Copy the console content to clipboard, including the escape sequences,
		Category(AnsiConsole.command.categoryid,Ansi Support Commands,null,true),
		WorkbenchHandlerServiceHandler("AnsiConsole.command.copy_with_escapes"),
		,,true),null),
	org.eclipse.ui.defaultAcceleratorConfiguration,
	org.eclipse.ui.console.ConsoleView,,,system)
Binding(CTRL+SHIFT+INSERT,
	ParameterizedCommand(Command(org.eclipse.ui.edit.text.toggleInsertMode,Toggle Insert Mode,
		Toggle insert mode,
		Category(org.eclipse.ui.category.edit,Edit,null,true),
		WorkbenchHandlerServiceHandler("org.eclipse.ui.edit.text.toggleInsertMode"),
		,,true),null),
	org.eclipse.ui.defaultAcceleratorConfiguration,
	org.eclipse.ui.textEditorScope,,,system)
```

So I would simply remove key mappings here until there is a better solution found.

I will push a PR in a moment that solves most of the issues I've observed and leave "keybinding / menu" issues to the contributor.

**Comment by iloveeclipse** [comment]
TODO's left for @mihnita after #58:

1) Understand where the keybinding conflicts with Java Stack Trace console are coming from - we have not just "one" console, but lot of different consoles, so if the keybinding issues appeared in one, they will also appear in other one for sure.
2) Move "copy with/without ANSI Escapes" actions from the "global" console menu to the concrete console page context menu (same menu where default Copy/Paste entries are, probably below of the "standard" copy/paste block). That's where the user would expect the actions, not in the global menu/toolbar.
3) Fix enablement of the options in the preference page: if "Enable ANSI Support" is unchecked, all options below should be disabled, and vice versa.
4) Whatever else I'm missing now.

**Comment by iloveeclipse** [comment]
Another point I see now: the preference constants must be better defined.
A product that wants disable ANSI support has to set now `org.eclipse.ui.console/booleanEnabled=false` which is for sure not OK, it must at least specify **what** is enabled in the console, not just some arbitrary "boolean".

I will patch that ASAP, because looking at the console matcher code in `org.eclipse.ui.internal.console.ansi.participants.AnsiConsoleStyleListener.findPositions(int, String)`  that is executed in UI thread, I will **for sure** disable this for our product where we might have **gazillions** of line printed in the console in the short time. Running that data over the matcher would most likely freeze UI.

 Surely we don't have a "usual" IDE use case, but product customization is important anyway.

**Comment by laeubi** [comment]
Thanks @iloveeclipse for taking care of this :+1:

**Comment by mihnita** [comment]
First, thank you for helping me make this better, and aligning it with Eclipse best practices.

Sorry for lack of quick response, but I am in US (looks like many of you are in Europe? :-)
And I can only work on this out of office.

But I will get to it asap.

I will (soon) send a PR for the issues that are easier to fix (Ansi => ANSI and such).
And I will summarize the others, so that you can prioritize them.

I'll try to get to all before the M3 deadline.
Btw, when is that? I know it is public on August 26, but when is the cutoff date for code changes?

Regards,
M.

**Comment by iloveeclipse** [comment]
> I will (soon) send a PR for the issues that are easier to fix (Ansi => ANSI and such).

Already done, please see comments above. Please pull from master before patching.

> Btw, when is that? I know it is public on August 26, but when is the cutoff date for code changes?

Github milestone says 19 August, means from Monday 15 August only regression fixes ahould be done, which would be OK here.

**Comment by mihnita** [comment]
I've create a PR 
"3. Fix enablement of the options in the preference page: if "Enable ANSI Support" is unchecked, all options below should be disabled, and vice versa."

**Comment by mihnita** [comment]
> 1. Understand where the keybinding conflicts with Java Stack Trace console are coming from - we have not just "one" console, but lot of different consoles, so if the keybinding issues appeared in one, they will also appear in other one for sure.

I have tried adding the Ctrl+Shift+C keybinding, and I don't see any conflict reported in the log.

If I go to Preferences -- General -- Keys and search for ANSI I see the Ctrl+Shit+C and Ctrl+Shit+Insert, with the "When" column being "In Console View"

If I search for "toggle comment" I see  Ctrl+Shit+C, but the entries "When" column is "C/C++ Editor", "Editing Java Source", "Editing Properties Files"

In both cases, if I click on the command with Ctrl+Shit+C bindings the "Conflicts" panel on the right stays empty.

Any idea how can I reproduce the situation you are seeing?

My testing was done on Windows, with the 2022‑09 M2 release, "Eclipse IDE for Java Developers", and replaced one single jar file (org.eclipse.ui.console_3.11.300.*.jar)

---

> 2. Move "copy with/without ANSI Escapes" actions from the "global" console menu to the concrete console page context menu (same menu where default Copy/Paste entries are, probably below of the "standard" copy/paste block). That's where the user would expect the actions, not in the global menu/toolbar.

I've spent some time on this one too.

I've managed to move all commands under the context menu, but they go between "Terminate/Disconnect All" and "Scroll Lock"

Q1. Would that be an acceptable location for "Enable / Disable ANSI Support"?

---

I am still trying (and don't succeed yet :-) to move the Copy options after the "standard copy"

I agree that is where it belongs, so I am still trying.

But I have a question here too: the plugin used to "hijack" the standard Copy command intentionally.

Most people want to copy what they see on screen (WYSIWYG).
That is the "Copy Text Without ANSI Escapes" option.

If left "un-hijacked" the standard Copy command does what StyledText does, and copies the escapes too.
That would be the same as "Copy Text With ANSI Escapes"

TLDR:
1. Since the standard command does either "Copy Text With ANSI Escapes", or "Copy Text Without ANSI Escapes", it does not make much sense to show all 3.
2. Most people want to copy WYSIWYG, which is "Copy Text Without ANSI Escapes". But that is not the default.

---

And last, a question about "ANSI Support Preferences..."

What do you think if I removed them completely?

There is already an option (at the end of the right-click popup menu) named "Preferences..."
That one opens the Preferences dialog, with Run/Debug > Console (2 pages)
I can modify it to show Run/Debug > Console > ANSI Support (3 pages)

It does not make much sense to have
```
ANSI Support Preferences...
Scroll Lock
Word Wrap
---
Preferences...
```
With the 2 "Preferences..." taking us to (almost) the same place.

---

Thank you very much,
Mihai

**Comment by iloveeclipse** [comment]
> I have tried adding the Ctrl+Shift+C keybinding, and I don't see any conflict reported in the log.

Open Console -> Java Stack Trace Console => kaboom.

> I've managed to move all commands under the context menu, but they go between "Terminate/Disconnect All" and "Scroll Lock"

OK for now, I think that is better as in the view options anyway.

> 2\. Most people want to copy WYSIWYG, which is "Copy Text Without ANSI Escapes". But that is not the default.

Sure. This will be probably more complicated, but I believe one can replace default "copy" command handler with own version, please create dedicate task for this, this thread is already too long to be useful.

> It does not make much sense to have
> 
> ```
> ANSI Support Preferences...
> ```

Yes, I see that too. I would say, we should replace `ANSI Support Preferences...` in the view menu with "default" `Preferences...` that opens root preference page for Console. And of course there is no need for that in the context menu.

**Comment by mihnita** [comment]
Created issue "ANSI Support: keybinding issues for Ctrl+Shift+C #65" and "ANSI Support: Cleanup context menu #66"

**Comment by iloveeclipse** [comment]
@mihnita : please update N&N screenshots: https://www.eclipse.org/eclipse/news/4.25/platform.php

**Comment by mihnita** [comment]
@iloveeclipse 

Would you consider the polishing done?

> TODO's left for @mihnita after https://github.com/eclipse-platform/eclipse.platform.debug/pull/58:

> 1. Understand where the keybinding conflicts with Java Stack Trace console are coming from - we have not just "one" console, but lot of different consoles, so if the keybinding issues appeared in one, they will also appear in other one for sure.

Keybinding removed. And we have a separate issue for it (issue #65), no need to track it twice.

> 2. Move "copy with/without ANSI Escapes" actions from the "global" console menu to the concrete console page context menu (same menu where default Copy/Paste entries are, probably below of the "standard" copy/paste block). That's where the user would expect the actions, not in the global menu/toolbar.

Removed completely.

> 3. Fix enablement of the options in the preference page: if "Enable ANSI Support" is unchecked, all options below should be disabled, and vice versa.

Done.

> 4. Whatever else I'm missing now.

Should we keep it open for this? :-)

> please update N&N screenshots: https://www.eclipse.org/eclipse/news/4.25/platform.php#debug-ansi-support

Done.

**Comment by iloveeclipse** [comment]
> > please update N&N screenshots: https://www.eclipse.org/eclipse/news/4.25/platform.php#debug-ansi-support
> 
> Done.

As commented on PR, the screenshots still show non existing buttons on console view and old preference page state.  Please update them.
Once this is done,  we are done here.
